File: suduko.cpp
#include <iostream>
#include <vector>

using namespace std;

class Sudoku {
public:
    Sudoku(string board_string);
    void solve();
    string board_to_string();

private:
    vector<vector<int>> board;
    bool solve_board(int row, int col);
    pair<int, int> next_position(int row, int col);
    bool valid_move(int row, int col, int num);
    vector<int> subgrid(int row, int col);
};

Sudoku::Sudoku(string board_string) {
    for (int i = 0; i < 81; i++) {
        int row = i / 9;
        int col = i % 9;
        if (board.size() <= row) {
            board.push_back(vector<int>());
        }
        board[row].push_back(board_string[i] - '0');
    }
}

void Sudoku::solve() {
    solve_board(0, 0);
}

bool Sudoku::solve_board(int row, int col) {
    if (row == 9) {
        return true;
    }

    pair<int, int> next_pos = next_position(row, col);
    int next_row = next_pos.first;
    int next_col = next_pos.second;

    if (board[row][col] != 0) {
        return solve_board(next_row, next_col);
    } else {
        for (int num = 1; num <= 9; num++) {
            if (valid_move(row, col, num)) {
                board[row][col] = num;
                if (solve_board(next_row, next_col)) {
                    return true;
                }
            }
        }
        board[row][col] = 0;
        return false;
    }
}

pair<int, int> Sudoku::next_position(int row, int col) {
    if (col == 8) {
        return { row + 1, 0 };
    } else {
        return { row, col + 1 };
    }
}

bool Sudoku::valid_move(int row, int col, int num) {
    for (int i = 0; i < 9; i++) {
        if (board[row][i] == num) {
            return false;
        }
        if (board[i][col] == num) {
            return false;
        }
    }
    for (int r : subgrid(row, col)) {
        if (r == num) {
            return false;
        }
    }
    return true;
}

vector<int> Sudoku::subgrid(int row, int col) {
    int subgrid_row = (row / 3) * 3;
    int subgrid_col = (col / 3) * 3;
    vector<int> result;
    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++) {
            result.push_back(board[subgrid_row + i][subgrid_col + j]);
        }
    }
    return result;
}

string Sudoku::board_to_string() {
    string result;
    for (int i = 0; i < 81; i++) {
        int row = i / 9;
        int col = i % 9;
        result += to_string(board[row][col]);
    }
    return result;
}

int main() {
    string board_string = "003020600900305001001806400008102900700000008006708200002609500800203009005010300";
    Sudoku sudoku(board_string);
   

File: suduko.rb
class Sudoku
  def initialize(board_string)
    @board = board_string.chars.map(&:to_i).each_slice(9).to_a
  end

  def solve
    solve_board
    to_s
  end

  def to_s
    @board.map { |row| row.join(' ') }.join("\n")
  end

  private

  def solve_board(row = 0, col = 0)
    return true if row == 9

    next_row, next_col = next_position(row, col)

    if @board[row][col] != 0
      solve_board(next_row, next_col)
    else
      (1..9).each do |num|
        if valid_move?(row, col, num)
          @board[row][col] = num
          return true if solve_board(next_row, next_col)
        end
      end
      @board[row][col] = 0
      false
    end
  end

  def next_position(row, col)
    if col == 8
      [row + 1, 0]
    else
      [row, col + 1]
    end
  end

  def valid_move?(row, col, num)
    return false if @board[row].include?(num)
    return false if @board.transpose[col].include?(num)
    return false if subgrid(row, col).include?(num)

    true
  end

  def subgrid(row, col)
    subgrid_row = (row / 3) * 3
    subgrid_col = (col / 3) * 3
    @board[subgrid_row, 3].map { |r| r[subgrid_col, 3] }.flatten
  end
end

board_string = '530070000600195000098000060800060003400803001700020006060000280000419005000080079'
sudoku = Sudoku.new(board_string)
puts sudoku.solve
class Board
  def initialize(board_string)
    @board = board_string.chars.map(&:to_i).each_slice(9).to_a
  end

  def to_s
    @board.map { |row| row.join(' ') }.join("\n")
  end

  def solved?
    Sudoku.new(board_string).solve == to_s
  end

  def solve
    @board = Sudoku.new(to_s).solve.chars.map(&:to_i).each_slice(9).to_a
  end

  def set(row, col, num)
    @board[row][col] = num
  end

  def get(row, col)
    @board[row][col]
  end

  private

  def board_string
    @board.flatten.join
  end
end
board_string = '530070000600195000098000060800060003400803001700020006060000280000419005000080079'
board = Board.new(board_string)

puts board.to_s

# Output:
# 5 3 0 0 7 0 0 0 0
# 6 0 0 1 9 5
